services/node_manager/src/main/java/com/futurewei/alcor/nodemanager/controller/NodeController.java

```diff
@@ -24,29 +24,49 @@
 import com.futurewei.alcor.nodemanager.service.NodeService;
 import com.futurewei.alcor.nodemanager.utils.NodeManagerConstant;
 import com.futurewei.alcor.common.utils.RestPreconditionsUtil;
+import com.futurewei.alcor.web.entity.node.BulkNodeInfoJson;
 import com.futurewei.alcor.web.json.annotation.FieldFilter;
 import io.swagger.annotations.ApiParam;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.http.HttpStatus;
 import org.springframework.web.bind.annotation.*;
 import org.springframework.web.multipart.MultipartFile;
+import com.futurewei.alcor.common.logging.Logger;
+import com.futurewei.alcor.common.logging.LoggerFactory;
 
 import javax.servlet.http.HttpServletRequest;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
+import java.util.logging.Level;
 
 import static org.springframework.web.bind.annotation.RequestMethod.*;
 
 @RestController
 public class NodeController {
+    private static final Logger LOG = LoggerFactory.getLogger();
 
     @Autowired
     private NodeService service;
 
     @Autowired
     private HttpServletRequest request;
 
+    @RequestMapping(
+            method = POST,
+            value = {"/bulk/nodes", "/v4/bulk/nodes"})
+    public List<NodeInfo> registerNodesInfoBulk(@RequestBody BulkNodeInfoJson resource) throws Exception {
+        if (resource == null) {
+            throw new ParameterNullOrEmptyException(NodeManagerConstant.NODE_EXCEPTION_FILE_EMPTY);
+        }
+        try {
+            return service.createNodeInfoBulk(resource.getNodeInfos());
+        } catch (Exception e) {
+            LOG.log(Level.SEVERE,e.getMessage());
+            throw e;
+        }
+    }
+
     @RequestMapping(
```

Comment:
Can you add a UT to cover the new /bulk/nodes API? Also, the /upload appears to be problematic. It will be nice to fix them all together.

Reply:
fixed in 72508 for adding UT

Reply:
Thanks. Can you please check the /upload UT and fix it if needed?

Reply:
fixed in 339c752